### DISCUSSION TITLE
Editable list, but can't add/remove items

### DISCUSSION THREAD

**Discussion by ajohnstone-ks** [MarimerLLC/csla#3540] [Questions] [answered]
I'm working on moving us to HttpProxy, and I've hit a problem I'm not sure what the recommended approach is.

Basically I have a BusinessListBase, which you can only fetch from the server.  The list is the UOW (so all changes to the child items are saved or discarded as a group), but you shouldn't be able to add or remove items from the list, just edit the existing instances.

So we added an IsReadOnly property, and overrode the InsertItem & RemoveItem to throw an exception if IsReadOnly.  We override onGetState/OnSetState to keep the value across the wire (its only set server side though).  

I guess with how WCF serialized things, it didn't use InsertItem to recreate the list.  But I think now that we're using HttpProxy, Csla does use InsertItem, and its after the OnSetState is called, so IsReadOnly is true and our exception gets thrown. 

Is there some other thing I need to do to keep this behavior, or will it just not work given how HttpProxy/MobileFormatter work?

I'm on Csla 4.5.2 and .Net Framework 4.7.2.

**Comment by swegele** [comment]
If there is a way you can detect the object is being transported across portal you could set the ReadOnly to false temporarily and then use the OnDeserialized event to set the readonly property back to true or something like that?

**Reply by ajohnstone-ks** [reply]
I don't know, that's why I'm here :-)  I think OnDeserialized though is only for the old school remoting & maybe wcf.  I can try overriding that and see if its called, its a good idea.

**Comment by danielmartind** [comment]
Can you set AllowAdd and AllowRemove to false?

**Reply by ajohnstone-ks** [reply]
Yes, but those are publicly settable too and don't actually prevent adding/removing.  I think WinForms databinding picks up on them but the list themselves I don't believe enforce anything there.

**Reply by ajohnstone-ks** [reply]
Although that said, the actual child BOs in the list only have [UpdateChild], but it would be nice to prevent a trip to the server before getting an exception.  So if I just have to abandon this I don't think it will be a big deal, at least for the list i'm looking at now.

**Comment by swegele** [comment]
I’ll see if I can poke at a little demo project today and get it to work.

**Reply by ajohnstone-ks** [reply]
I'm going to keep playing with it too, but ultimately if I can't do it I think we can live without it.

**Reply by ajohnstone-ks** [reply]
So I did override OnDeserialized, and its not called at all.  I'm going to move on to addressing other issues and come back to this. I appreciate your response!

**Reply by rockfordlhotka** [reply, marked as answer]
I believe you want to implement [ISerializationNotification](https://github.com/MarimerLLC/csla/blob/main/Source/Csla/Serialization/Mobile/ISerializationNotification.cs#L16) to get a `Deserialized` method invocation when deserialization of the instance (and its children) is complete.

https://github.com/MarimerLLC/csla/blob/969bee634bd5f181ed18ef8358dba2b6a89d0d04/Source/Csla/Serialization/Mobile/MobileFormatter.cs#L312

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka Ahh, that looks promising!  I will try that.

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka Thanks, that did the trick!  I store the flag in a temporary variable and then in Deserialized I set the property using the temp variable.
